### PR TITLE
Refactor/GitHub workflow

### PR DESCRIPTION
**Pull request by harrysarson**
Opening this PR to check the `on: pull_request` workflow.

### CODE REVIEW

**Comment by josdejong** [comment]
I've added the browserstack keys to the secrets section of github, all works like a charm. I really like the neat separation in files to run on `push` vs `pull_request`. 

Should the build.yaml file also run on `pull_request`?

**Comment by harrysarson** [comment]
> Should the build.yaml file also run on pull_request?

Yes it should.

File: package.json [hunk 1]
```diff
@@ -121,7 +121,6 @@
     "validate:ascii": "gulp --gulpfile gulpfile.cjs validate:ascii",
     "test": "npm run test:src && npm run lint",
     "test:src": "mocha test/unit-tests --recursive --forbid-only",
-    "test:src:transpile": "mocha test/unit-tests --require @babel/register --recursive --forbid-only",
```
[the remark this answers is not in the document]
Comment (josdejong) [reply package.json:124]: I created this script temporarily to compare the duration of `test:src` with and without babel transpilation. I should have cleaned up but it sneaked in the develop branch. So now I'm trying to sneak it out again ;)

**Comment by josdejong** [comment]
Thanks for the update Harry. 

I'm wondering, it looks like all tasks of `build.yml` are executed twice: once for push and once for pull_request. Any idea how we can prevent that from happening?

And I guess the "continuous-integration/travis-ci" check is a left over and will be gone after we've merged all.

**Comment by harrysarson** [comment]
I think that happens because this is both a branch (in the mathjs repo) and a pull request (against the mathjs repo). This would not happen if someone makes a pull request from a fork this workflow would not run twice.

I am not sure if we can stop it running twice for branches that are also pull requests.

**Comment by josdejong** [comment]
Hm. So maybe when you update a PR, push is *also* triggered? Would it be enough to just trigger build.yml on `push` after all then, and not both?

**Comment by harrysarson** [comment]
> Would it be enough to just trigger build.yml on push after all then

Not if we want the workflow to run when someone makes a pull request from a fork

**Comment by josdejong** [comment]
Hm, I guess we have too keep it like this though. It's similar to travis where we also had this, right? (only you didn't see this in a single list with all tasks).

So I think we can merge this PR, right @harrysarson ?

**Comment by harrysarson** [comment]
> Hm, I guess we have too keep it like this though. 

Currently:

```
on: [push, pull_request]
```

We could:

```
on:
  push:
    branches:
      - develop
      - master
    tags:  
  pull_request:
```

because any other branch would end up being a pull_request so the build job would get run then.

> It's similar to travis where we also had this, right? (only you didn't see this in a single list with all tasks).

Yes, although running less CI jobs is always a good thing if we can do it!

**Comment by josdejong** [comment]
I think the `branches:` configuration makes sense, we can try it out. What do you think?

File: .github/workflows/browserstack.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+name: Browser testing on browser stack
+
+on: push
```
Comment (harrysarson) [inline .github/workflows/browserstack.yaml:3]:
```suggestion
on:
  push:
    branches:
      - develop
      - master
    tags:  
  pull_request:
```

**Comment by harrysarson** [comment]
Oops, I changed the wrong file! Let me fix next time I get to my laptop 😊

**Comment by harrysarson** [comment]
Do check when this is merged into dev that the build job runs :)

Is it intentional that we do not run CI on nodejs 10?

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (harrysarson) [reply package.json:124]: sorry for breaking your cover!

**Comment by josdejong** [comment]
> 
> 
> Do check when this is merged into dev that the build job runs :)
> 
> Is it intentional that we do not run CI on nodejs 10?

Ahhh, wait a second. In the old Travis script we have this part:

https://github.com/josdejong/mathjs/blob/59aacb4c04dac169572c8fa5d3d2961e890f41ca/.travis.yml#L14-L16

So I think my former me made this `test:src:transpile` script specially for nodejs 10, which is missing some of the newer JavaScript features and requires transpilation before you can run the tests. Argh. So I guess we'll have to add this script again and a Github workflow for it, until end-of-life of nodejs 10, which is coming April 2021.

**Comment by harrysarson** [comment]
Take a look! We now run special script on node 10. I also tidied the scripts up nicely.

**Comment by josdejong** [comment]
Looks good, nicely solved with the node 10 script, having all logic for this in one place 👍 .

Ready to merge I think?

**Comment by harrysarson** [comment]
fixes #2024

**Comment by harrysarson** [comment]
ready to merge I think

**Comment by josdejong** [comment]
👍 thanks again!

**Comment by harrysarson** [comment]
Is good to be back! Merry Christmas Josh, till next time.

**Comment by josdejong** [comment]
It works like a charm, this is awesome 😎 

Happy holidays Harry!

**Pull request merged**